### PR TITLE
[SPARK-16927] [SPARK-16923] Override task properties at dispatcher.

### PR DESCRIPTION
**Pull request by mgummelt**
## What changes were proposed in this pull request?
- enable setting default properties for all jobs submitted through the dispatcher [SPARK-16927]
- remove duplication of conf vars on cluster submitted jobs [SPARK-16923](this is a small fix, so I'm including in the same PR)
## How was this patch tested?

mesos/spark integration test suite
manual testing

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 1]
```diff
@@ -358,14 +358,12 @@ private[spark] class MesosClusterScheduler(
         .orElse(desc.command.environment.get("SPARK_EXECUTOR_URI"))
   }
 
-  private def getDriverEnvironment(desc: MesosDriverDescription): Environment = {
-    val env = {
-      val executorOpts = desc.conf.getAll.map { case (k, v) => s"-D$k=$v" }.mkString(" ")
-      val executorEnv = Map("SPARK_EXECUTOR_OPTS" -> executorOpts)
```
Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Not only are these variables already set in `generateCmdOption`, but SPARK_EXECUTOR_OPTS has no effect on the driver process

**Comment by mgummelt** [comment]
@skonto Can you take a look at this?

**Comment by mgummelt** [comment]
retest please

**Comment by skonto** [comment]
I will have a look shortly

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 1 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: From what I understand SPARK_EXECUTOR_OPTS is not used because it is actually reconstructed at the driver side
[here:](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala#L180)  from spark.executor.extraJavaOptions.
I dont see that property set at the Dispacther side.

Two things here:
I would prefer the process of reconstructing the env to be transparent to the user, the same for client and cluster mode, no need for special prefixes. 
Secondly, we need to make sure there is continuity with all properties we pass from process to process.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 2]
```diff
@@ -463,9 +455,20 @@ private[spark] class MesosClusterScheduler(
...
     }
-    desc.conf.getAll
+
+    // --conf
+    val replicatedOptionsBlacklist = Set(
+      "spark.jars", // Avoids duplicate classes in classpath
+      "spark.submit.deployMode", // this would be set to `cluster`, but we need client
+      "spark.master" // this contains the address of the dispatcher, not master
+    )
+    val defaultConf = conf.getAllWithPrefix("spark.mesos.dispatcher.driverProperty.").toMap
```
Comment (skonto) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Property suffix seems verbose to me.

File: docs/running-on-mesos.md [hunk 3]
```diff
@@ -467,7 +467,19 @@ See the [configuration page](configuration.html) for information on Spark config
...
     If unset it will point to Spark's internal web UI.
   </td>
+  </tr>
+<tr>
+  <td><code>spark.mesos.dispatcher.driverProperty.[PropertyName]</code></td>
+  <td><code>(none)</code></td>
+  <td>
+    Set default properties for drivers submitted through the
+    dispatcher.  For example,
+    spark.mesos.dispatcher.driverProperty.spark.executor.memory=32g
```
Comment (skonto) [inline docs/running-on-mesos.md]: From a UX perspective I would like to set spark.executor.memory=32g without caring about where I am running on.... test env, real cluster client or cluster mode...

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 4]
```diff
@@ -463,9 +455,20 @@ private[spark] class MesosClusterScheduler(
...
+    val driverConf = desc.conf.getAll
       .filter { case (key, _) => !replicatedOptionsBlacklist.contains(key) }
-      .foreach { case (key, value) => options ++= Seq("--conf", s"$key=${shellEscape(value)}") }
+      .toMap
+    (defaultConf ++ driverConf).foreach { case (key, value) =>
```
Comment (skonto) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: what if I have the same property assigned both at driverConf and defaultConf should not be the case that the former takes precedence?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 1 above)
Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: I've reread this a few times, and still can't understand it.  Can you rephrase this comment?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 2 above)
Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: That's just as verbose

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 4 above)
Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: It does: http://stackoverflow.com/questions/11742451/concatenate-two-immutable-maps-which-elements-are-preferred

Comment (skonto) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: The answer there does not have a clear reference assume that the behavior is consistent.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 2 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: I know just a bit more explicit about the semantics

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 1 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: I rephrased it...

Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Thanks.

The one way to pass configs through.  The only thing I would change in your comment is "to all executors", because most spark properties aren't set on executors.  They're set on drivers, and often influence how the drivers launch the executors.

That single way is now in `generateCmdOption`, which translates `desc.conf` into `--conf` parameters for the driver.  Removing this is removing a second, redundant way of setting the same configs.

There is another way users can pass configs through, which is through the deprecated `SPARK_SUBMIT_OPTS` env var.  That's passed through just as all env vars are passed through, which is in this method (`desc.command.environment`)

**Comment by skonto** [comment]
Ok LGTM other than the name of the setting for which I dont have a much better alternative.  
@srowen could you check and merge pls?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 2 above)
Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: I agree, except I don't think there's any need to make a separate `driver` level in the hierarchy.  I changed it to `driverDefault`.

**Comment by srowen** [comment]
Merged to master
